### ISSUE TITLE
Standardize MBQL clause "options"

### ISSUE THREAD

**Issue by tlrobinson** [metabase/metabase#11953]
We’re looking to add “named” filters like we have for aggregations and need somewhere in MBQL store them, like we have for `:aggregation-options`. Eventually we may want to add names for every clause in the notebook.

IMHO the `:aggregation-options` clause is a bit annoying to work with, since we need to always check if an aggregation is wrapped in `aggregation-options` first. `metabase-lib` on the FE helps with that a bit but it still feels kind of wrong.

Also, filters are a bit different than aggregations, since there's just a single `filter` in the query, and we just `:and` multiple filters. So we'd need to do something like:

```edn
{
  :source-table 1,
  :filter
    [:and
      [:filter-options [:or [:= [:field-id 1] 42] [:= [:field-id 2] 84]] {:display-name "foo" }]
      [:filter-options [:= [:field-id 3] 1234]] {:display-name "bar" }]]
}
```

Rather than adding `:filter-options`, one idea i had was to specify "if the last item in any MBQL clause is a map, that is the options for that clause" and put `:display-name` there. It may also be useful for things like a unique `:key` as described in solution 6 here: #11829. Any clause could have a `:display-name`, and optionally may have clause-specific options.

i.e. instead of above we'd do

```edn
{
  :source-table 1,
  :filter
    [:and
      [:or [:= [:field-id 1] 42] [:= [:field-id 2] 84]] {:display-name "foo" }]
      [:= [:field-id 3] 1234 {:display-name "bar" }]]
}
```

We already have options for a number of clauses:

1. `:contains`/`:starts-with`/`:ends-with`/`:does-not-contain`'s `:case-sensitive`
2. `:time-interval`'s `:include-current`
3. `:aggregation-options` itself
4. `:case`'s `:default` 

I don't believe we use maps in MBQL s-expression style clauses for any purpose other an options, so this wouldn't conflict with any existing syntax, however optional and variable number of arguments might make this a little more annoying to deal with.

I think ideally the "options" map should replace any optional arguments like `:relative-datetime`'s `:unit`, `:binning-strategy` parameters, etc. I'm a fan of [this style](https://deno.land/std/style_guide.md#exported-functions-max-2-args-put-the-rest-into-an-options-object):

> A function that is part of the public API takes 0-2 required arguments, plus (if necessary) an options object (so max 3 total).

The remaining issue is clauses with variable number of arguments like `:=`, `:+`, `:and`, etc. For `:case` we made the first argument an array of case clauses rather than having the `:case` clause itself have a variable number of arguments, but doing `[:+ [a b c]]` instead of `[:+ a b c]` seems kind of lame. Maybe clauses with variable number of arguments needs just make sure they ignore the options map at the end? `[:+ a b c {:display-name "foo"}]`

**Comment by camsaul** [comment]
The whole s-expression syntax is a little annoying IMO we would probably have easier lives if we just used maps for everything

**Comment by camsaul** [comment]
But on the other hand I think something like changing

```clj
[:+ [:field-id 1] 100]
```

to

```clj
{:type :+, :args [{:type :field-id, :id 1} 100]}
```

is actually even more annoying so maybe that's not something we want to do. 🤷‍♂ 

So yeah I guess we could do something like accept maps at the end of things it would require reworking some stuff but should be doable.

**Comment by tlrobinson** [comment]
Yeah, I think s-expressions with only required arguments and an optional options map would be a good compromise. 

I think variable number of arguments is the main open question for me.

**Comment by camsaul** [comment]
If the last arg is the only thing that can be a map I don't think that should be too hard to deal with

**Comment by sbelak** [comment]
I like this as we can then also somewhat decouple the logic and eg. handle the presentation stuff in a separate pass.

**Comment by walterl** [comment]
It feels like there's space for ambiguity with rules like "if the last arg is a map, it's options". We could perhaps be a little bit more explicit about it, and wrap the options map in a vector with a marker:

```clojure
{
  :source-table 1,
  :filter
    [:and
      [:or [:= [:field-id 1] 42] [:= [:field-id 2] 84]] [:options {:display-name "foo" }]]
      [:= [:field-id 3] 1234 [:options {:display-name "bar" }]]]
}
```

Is the verbosity worth the added clarity?

**Comment by camsaul** [comment]
Implemented by MLv2 #28689
